fuzzing/fuzz/fuzz_targets/fuzz_target_2.rs

```diff
@@ -0,0 +1,310 @@
+//! An advanced fuzz test.
+//!
+//! This demonstrates use of the `SorabanArbitrary` trait,
+//! and the advancement of time.
+
+#![no_main]
+
+use libfuzzer_sys::fuzz_target;
+use soroban_fuzzing_contract::*;
+use soroban_ledger_snapshot::LedgerSnapshot;
+use soroban_sdk::arbitrary::arbitrary::{self, Arbitrary, Unstructured};
+use soroban_sdk::arbitrary::{fuzz_catch_panic, SorobanArbitrary};
+use soroban_sdk::testutils::{Address as _, LedgerInfo};
+use soroban_sdk::token::AdminClient as TokenAdminClient;
+use soroban_sdk::token::Client as TokenClient;
+use soroban_sdk::xdr::ScAddress;
+use soroban_sdk::{Address, Env, FromVal, Vec};
+use std::vec::Vec as RustVec;
+
+const NUM_ADDRESSES: usize = 2;
+
+#[derive(Arbitrary, Debug)]
+struct Input {
+    addresses: [<Address as SorobanArbitrary>::Prototype; NUM_ADDRESSES],
+    #[arbitrary(with = |u: &mut Unstructured| u.int_in_range(0..=i128::MAX))]
+    token_mint: i128,
+    steps: RustVec<Step>,
+}
+
+#[derive(Arbitrary, Debug)]
+struct Step {
+    #[arbitrary(with = |u: &mut Unstructured| u.int_in_range(1..=u64::MAX))]
+    advance_time: u64,
+    command: Command,
+}
+
+#[derive(Arbitrary, Debug)]
+enum Command {
+    Deposit(DepositCommand),
+    Claim(ClaimCommand),
+}
+
+#[derive(Arbitrary, Debug)]
+struct DepositCommand {
+    #[arbitrary(with = |u: &mut Unstructured| u.int_in_range(0..=NUM_ADDRESSES - 1))]
+    depositor_index: usize,
+    amount: i128,
+    // This is an ugly way to get a vector of integers in range
+    #[arbitrary(with = |u: &mut Unstructured| {
+        u.arbitrary_len::<usize>().map(|len| {
+            (0..len).map(|_| {
+                u.int_in_range(0..=NUM_ADDRESSES - 1)
+            }).collect::<Result<RustVec<usize>, _>>()
+        }).and_then(|inner_result| inner_result)
+    })]
+    claimant_indexes: RustVec<usize>,
+    time_bound: <TimeBound as SorobanArbitrary>::Prototype,
+}
+
+#[derive(Arbitrary, Debug)]
+struct ClaimCommand {
+    #[arbitrary(with = |u: &mut Unstructured| u.int_in_range(0 ..= NUM_ADDRESSES - 1))]
+    claimant_index: usize,
+    amount: i128,
+}
+
+fuzz_target!(|input: Input| {
+    let (config, mut prev_env) = Config::setup(input);
+
+    for step in &config.input.steps {
+        // Advance time and create a new env from snapshot.
+        let curr_env = {
+            let mut snapshot = prev_env.to_snapshot();
+            snapshot.sequence_number += 1;
+            snapshot.timestamp = snapshot.timestamp.saturating_add(step.advance_time);
+            let env = Env::from_snapshot(snapshot);
+            env.budget().reset_unlimited();
+            env
+        };
+
+        step.command.exec(&config, &curr_env);
+        assert_invariants(&config, &prev_env, &curr_env);
+
+        prev_env = curr_env;
+    }
+});
```

Comment:
This code uses soroban_ledger_snapshot to save and restore a test ledger to simulate advancing time. I am unsure if this is the best way to do this. Please review closely.

Reply:
If all you want to do is advance the ledger seq and timestamp, you can do something like this (and add code to bump the timestamp) - 

```rust
e.ledger().with_mut(|li| li.sequence_number = li.sequence_number + 1);
```

Reply:
I have done that in the past, but I don't think the semantics are quite what I want. It's not just the seq/timestamp that I want to change, but to simulate multiple transactions, and this approach leaves the Env intact, where multiple transactions would destroy and recreate the Env. One difference in the two approaches that I am aware of is that the budget will not be reset by just mutating the ledger in this way.

Reply:
Of course, these examples turn the budget off explicitly, so that exact point is moot in this case.

Reply:
Oh interesting. This might be the best approach then.